src/unwinder/arch/loongarch64.rs

```diff
@@ -0,0 +1,228 @@
+use core::{fmt, ops};
+use gimli::{LoongArch, Register};
+
+use super::maybe_cfi;
+
+pub const MAX_REG_RULES: usize = 74;
+
+#[repr(C)]
+#[derive(Clone, Default)]
+pub struct Context {
+    pub gp: [usize; 32],
+    pub fp: [usize; 32],
+}
+
+impl fmt::Debug for Context {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut fmt = fmt.debug_struct("Context");
+        for i in 0..=31 {
+            fmt.field(
+                LoongArch::register_name(Register(i as _)).unwrap(),
+                &format_args!("{:#x}", self.gp[i]),
+            );
+        }
+        for i in 0..=31 {
+            fmt.field(
+                LoongArch::register_name(Register((i + 32) as _)).unwrap(),
+                &format_args!("{:#x}", self.fp[i]),
+            );
+        }
+        fmt.finish()
+    }
+}
+
+impl ops::Index<Register> for Context {
+    type Output = usize;
+
+    fn index(&self, reg: Register) -> &usize {
+        match reg {
+            Register(0..=31) => &self.gp[reg.0 as usize],
+            Register(32..=63) => &self.fp[(reg.0 - 32) as usize],
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl ops::IndexMut<gimli::Register> for Context {
+    fn index_mut(&mut self, reg: Register) -> &mut usize {
+        match reg {
+            Register(0..=31) => &mut self.gp[reg.0 as usize],
+            Register(32..=63) => &mut self.fp[(reg.0 - 32) as usize],
+            _ => unimplemented!(),
+        }
+    }
+}
+
+macro_rules! save_regs {
+    (gp$(, $fp:ident)?) => {
+        core::arch::naked_asm!(
+            maybe_cfi!(".cfi_startproc"),
+            "
+            move $r12, $r3
+            addi.d $r3, $r3, -0x210
+            ",
+            maybe_cfi!(".cfi_def_cfa_offset 0x210"),
+            "
+            st.d $r1, $r3, 0x200
+            ",
+            maybe_cfi!(".cfi_offset 1, -16"), // ra
+            "
+            st.d $r0, $r3, 0x0 // zero
+            st.d $r1, $r3, 0x8 // ra
+            st.d $r2, $r3, 0x10 // tp
+            // sp is saved later
```

Comment:
Any reason to defer the saving here? The value is in `$t0` now, so you can avoid recalculating it later. See how rv64 code does this.